comRamona/assignment3/WordSearch.java

```diff
@@ -0,0 +1,88 @@
+package assignment3;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Given a grid of letters and a dictionary, find all the words from the dictionary that can be
+ * formed in the grid.
+ * The rules for forming a word:
+ * - You can start at any position.
+ * - You can move to one of the 8 adjacent cells (horizontally/vertically/diagonally).
+ * - You can't visit the same cell twice in the same word.
+ */
+public class WordSearch {
+    private Dictionary dict;
+    private char[][] grid;
+    private int rows;
+    private int cols;
+
+    /**
+     * @param dict dictionary object that can be queried for valid words and prefixes
+     * @param grid input grid
+     */
+    public WordSearch(Dictionary dict, char[][] grid) {
+        this.dict = dict;
+        this.grid = grid;
+        rows = grid.length;
+        cols = grid[0].length;
+    }
+
+    /**
+     * @return set of words found in grid
+     */
+    public Set<String> findWords() {
+        Set<String> results = new HashSet<>();
+        for (int i = 0; i < rows; i++) {
+            for (int j = 0; j < cols; j++) {
+                boolean[][] visited = new boolean[rows][cols];
```

Comment:
Creating a whole boolean array each time is a memory churn. Can we re-use the same array?